sadx-input-mod/KeyboardMouse.cpp

```diff
@@ -2,22 +2,50 @@
 
 #include "minmax.h"
 #include "KeyboardMouse.h"
+#include "SADXKeyboard.h"
 
 DataPointer(HWND, hWnd, 0x3D0FD30);
 
 ControllerData KeyboardMouse::pad           = {};
 float          KeyboardMouse::normalized_l_ = 0.0f;
 float          KeyboardMouse::normalized_r_ = 0.0f;
 bool           KeyboardMouse::mouse_active  = false;
+bool           KeyboardMouse::wheel_active  = false;
 bool           KeyboardMouse::left_button   = false;
 bool           KeyboardMouse::right_button  = false;
 bool           KeyboardMouse::half_press    = false;
+bool           KeyboardMouse::e_held	    = false;
 NJS_POINT2I    KeyboardMouse::cursor        = {};
 KeyboardStick  KeyboardMouse::sticks[2]     = {};
 Sint16         KeyboardMouse::mouse_x       = 0;
 Sint16         KeyboardMouse::mouse_y       = 0;
+Sint16         KeyboardMouse::wheel_delta	= 0;
 WNDPROC        KeyboardMouse::lpPrevWndFunc = nullptr;
 
+void KeyboardMouse::clear_sadx_keys(bool force)
+{
+	for (int i = 0; i < LengthOfArray(SADX2004Keys); i++)
+	{
+		KeyboardKeys[SADX2004Keys[i].SADX2004Code].old = KeyboardKeys[SADX2004Keys[i].SADX2004Code].held;
+		KeyboardKeys[SADX2004Keys[i].SADX2004Code].pressed = 0;
+		if (force) KeyboardKeys[SADX2004Keys[i].SADX2004Code].held = 0;
+	}
+}
+
+void KeyboardMouse::update_sadx_key(Uint32 key, bool down)
+{
+	for (int i = 0; i < LengthOfArray(SADX2004Keys); i++)
+	{
+		if (key == SADX2004Keys[i].WindowsCode)
+		{
+			if (KeyboardKeys[SADX2004Keys[i].SADX2004Code].old != down) KeyboardKeys[SADX2004Keys[i].SADX2004Code].pressed = down;
+			KeyboardKeys[SADX2004Keys[i].SADX2004Code].old = KeyboardKeys[SADX2004Keys[i].SADX2004Code].held;
+			KeyboardKeys[SADX2004Keys[i].SADX2004Code].held = down;
+			return;
+		}
+	}
```

Comment:
Given the way it is used, I would suggest that `SADX2004Keys` be changed to `std::map`. Its definition would look something like:
```cpp
std::map<int, int> SADX2004Keys
```

And this entire for loop can be completely removed. e.g.:
```cpp
auto it = SADX2004Keys.find(key);

// no equivalent windows -> sadx mapping available
if (it == SADX2004Keys.end())
{
    return;
}

auto& keyboard_key = KeyboardKeys[it->second];

if (keyboard_key.old != down) keyboard_key.pressed = down;
keyboard_key.old = keyboard_key.held;
keyboard_key.held = down;
```